### ISSUE TITLE
Add git tags for frog versions

### ISSUE THREAD

**Issue by eestrada** [greghendershott/frog#184]
I would open this as a pull request, but it isn't possible to make pull requests for `git` tags on GitHub.

I figure tags make it easier to diff between versions, do regression testing, or install historical versions of frog (e.g. `raco pkg install https://github.com/greghendershott/frog.git#v0.25`).

In order to integrate this, if you choose to do so, you will need to `git fetch --tags` or `git pull --tags` directly from my GitHub repo to your local repo, then `git push origin --tags` to this repo to actually get the tags migrated. As I said, there isn't another way to merge in tags for GitHub that I know of.

Here are the relevant links for my fork:

* Repo: https://github.com/eestrada/frog.git
* Tags: https://github.com/eestrada/frog/tags
* Releases: https://github.com/eestrada/frog/releases (still not sure why/how github differentiates these from regular old annotated tags)

**Comment by greghendershott** [comment]
I don't know. My first, gut reaction is to avoid this. Working out why I feel that way:

- Although I appreciate you doing the work, now, this will be an extra chore going forward. And just for Frog? Why not also transitively for its dependencies (e.g. markdown, rackjure)? What about deps I don't control?

- This seems to go against the idea of storing data once -- put a value in `version` in `info.rkt`, and also put the (hopefully) same value in a git tag naming (hopefully) the correct commit of `info.rkt`. ("Hopefully" = it's not just the extra work of saying something twice; it's the chance of saying different things.)

- Racket packages use [checksums](http://docs.racket-lang.org/pkg/Package_Concepts.html#%28tech._checksum%29) to identify releases, not [versions](http://docs.racket-lang.org/pkg/Package_Concepts.html#%28tech._version%29):
  > A version is intended to reflect available features of a package, and should not be confused with different releases of a package as indicated by the checksum.

In other words I think the way to identify different points in Frog's history is through git commit checksums. I'm not sure I see the value of using tags to make aliases for some of the commits.

The use of `version`s is about dependencies. For instance the version of the markdown package matters, because it matters in Frog's `deps`.  But the version of Frog itself doesn't matter, unless it were a dep of some other package.

---

Maybe this would be a useful discussion to have (on the mailing list?) about Racket packages generally?

Maybe there could be a `version` -> `checksum` utility that works for _any_ package? I don't think this already exists in `raco pkg` but maybe you could submit a PR to add that?

**Comment by tger** [comment]
For what it's worth, I have the plugin API in the pipe, and when/if that's merged, plugin packages should probably depend on specific Frog version(s) to make the API is available.

**Comment by greghendershott** [comment]
Editorial warning. :smile:

Generally speaking, I think checking for version numbers is often a kind of "XY question" -- a proxy for what we really care about. Usually, we care whether _specific capabilities_ are available. A good system will let us ask _that_ question directly. (And if the capability is present, great, we use it. If not, we try to degrade gracefully, or, provide an error like "please install the latest foo". (Not version x of foo. Just, "the latest".))

In COM this is `QueryInterface`. In Emacs Lisp, `fboundbp`. In Racket, `dynamic-require`.

For example I [do this in racket-mode](https://github.com/greghendershott/racket-mode/blob/5279cda4a9385130cf7cc97bbdd33260deb0720d/instrument.rkt#L33-L40). I don't ask what version of Racket is running, I ask for the specific thing I need.

When possible I think this is a much better way for us devs and for users, both.

**Comment by tger** [comment]
I totally agree. However, in the case Racket packages `info.rkt` files the only supported attributes to check for are version and platform. Anyway, at worst a plugin would just not be picked up by Frog, you'll see what I have in mind once I send the pull request.

**Comment by eestrada** [comment]
@greghendershott This is all good food for thought. After chewing on it for bit, I agree with pretty much everything you said. 

I agree that the best way forward is probably to bring this up on the Racket mailing list and see if there is someway to get versions mapped to checksum releases working in `raco pkg` and `pkgs.racket-lang.org`. Or maybe I should just code it up, create a pull request, and have the discussion of the merit of this live there.

**Comment by greghendershott** [comment]
I'm closing because this is two years old and in my opinion not actionable now.
